### ISSUE TITLE
Parse not definded (Online Cup)

### ISSUE THREAD

**Issue by SJang1** [rishubil/TrickyTowersUtils#16]
```
Traceback (most recent call last):
  File "observer.py", line 105, in <module>
  File "memory_parser.py", line 231, in parse
  File "memory_parser.py", line 150, in parse_net_player_startups
  File "models.py", line 289, in value
  File "models.py", line 289, in <listcomp>
  File "site-packages\pymem\__init__.py", line 568, in read_ushort
pymem.exception.MemoryReadError: Could not read memory at: 113250304, length: 2 - GetLastError: 299

The token does not have the specified privilege. 

2019-03-19 18:48:30,223 - pymem - DEBUG - Process 22508 is being debugged

--- Parse at 1552988910224 ---
[!] Cannot get GameStateController = Cannot access memory: 0xc, (0x1f6964->0x3a0->0x6a8->0xc)
[?] game_info.game_mode is not defined
[?] game_info.game_type is not defined
[?] players[0].username is not defined
[?] players[0].id is not defined
[?] players[0].is_online is not defined
[?] players[1].username is not defined
[?] players[1].id is not defined
[?] players[1].is_online is not defined
[?] players[2].username is not defined
[?] players[2].id is not defined
[?] players[2].is_online is not defined
[?] players[3].username is not defined
[?] players[3].id is not defined
[?] players[3].is_online is not defined
[?] players[4].username is not defined
[?] players[4].id is not defined
[?] players[4].is_online is not defined
[?] players[5].username is not defined
[?] players[5].id is not defined
[?] players[5].is_online is not defined
[?] players[6].username is not defined
[?] players[6].id is not defined
[?] players[6].is_online is not defined
[?] players[7].username is not defined
[?] players[7].id is not defined
[?] players[7].is_online is not defined
[?] players[8].username is not defined
[?] players[8].id is not defined
[?] players[8].is_online is not defined
[?] players[9].username is not defined
[?] players[9].id is not defined
[?] players[9].is_online is not defined
[?] players[10].username is not defined
[?] players[10].id is not defined
[?] players[10].is_online is not defined
[?] players[11].username is not defined
[?] players[11].id is not defined
[?] players[11].is_online is not defined
[?] players[12].username is not defined
[?] players[12].id is not defined
[?] players[12].is_online is not defined
[?] players[13].username is not defined
[?] players[13].id is not defined
[?] players[13].is_online is not defined
[?] players[14].username is not defined
[?] players[14].id is not defined
[?] players[14].is_online is not defined
[?] players[15].username is not defined
[?] players[15].id is not defined
[?] players[15].is_online is not defined
[?] players[16].username is not defined
[?] players[16].id is not defined
[!] Tricky Towers is may not running: MemoryReadError('Could not read memory at: 113250304, length: 2 - GetLastError: 299')
Traceback (most recent call last):
  File "site-packages\pymem\__init__.py", line 566, in read_ushort
  File "site-packages\pymem\memory.py", line 238, in read_ushort
  File "site-packages\pymem\memory.py", line 105, in read_bytes
pymem.exception.WinAPIError: Windows api error, error_code: 299
```

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "observer.py", line 105, in <module>
  File "memory_parser.py", line 231, in parse
  File "memory_parser.py", line 150, in parse_net_player_startups
  File "models.py", line 289, in value
  File "models.py", line 289, in <listcomp>
  File "site-packages\pymem\__init__.py", line 568, in read_ushort
pymem.exception.MemoryReadError: Could not read memory at: 113250304, length: 2 - GetLastError: 299

The token does not have the specified privilege. 

2019-03-19 18:48:34,761 - pymem - DEBUG - Process 22508 is being debugged

--- Parse at 1552988914762 ---
[!] Cannot get GameStateController = Cannot access memory: 0xc, (0x1f6964->0x3a0->0x6a8->0xc)
[?] game_info.game_mode is not defined
[?] game_info.game_type is not defined
[?] players[0].username is not defined
[?] players[0].id is not defined
[?] players[0].is_online is not defined
[?] players[1].username is not defined
[?] players[1].id is not defined
[?] players[1].is_online is not defined
[?] players[2].username is not defined
[?] players[2].id is not defined
[?] players[2].is_online is not defined
[?] players[3].username is not defined
[?] players[3].id is not defined
[?] players[3].is_online is not defined
[?] players[4].username is not defined
[?] players[4].id is not defined
[?] players[4].is_online is not defined
[?] players[5].username is not defined
[?] players[5].id is not defined
[?] players[5].is_online is not defined
[?] players[6].username is not defined
[?] players[6].id is not defined
[?] players[6].is_online is not defined
[?] players[7].username is not defined
[?] players[7].id is not defined
[?] players[7].is_online is not defined
[?] players[8].username is not defined
[?] players[8].id is not defined
[?] players[8].is_online is not defined
[?] players[9].username is not defined
[?] players[9].id is not defined
[?] players[9].is_online is not defined
[?] players[10].username is not defined
[?] players[10].id is not defined
[?] players[10].is_online is not defined
[?] players[11].username is not defined
[?] players[11].id is not defined
[?] players[11].is_online is not defined
[?] players[12].username is not defined
[?] players[12].id is not defined
[?] players[12].is_online is not defined
[?] players[13].username is not defined
[?] players[13].id is not defined
[?] players[13].is_online is not defined
[?] players[14].username is not defined
[?] players[14].id is not defined
[?] players[14].is_online is not defined
[?] players[15].username is not defined
[?] players[15].id is not defined
[?] players[15].is_online is not defined
[?] players[16].username is not defined

(Repete on observer)
& No name came out
```

![image](https://user-images.githubusercontent.com/29011440/54596381-debcf580-4a77-11e9-84e3-d5117f9869a9.png)
(localhost, playing game)

Running GUI.

**Comment by SJang1** [comment]
For server : 
received message: {'game_info': {'is_playing': False, 'is_finished': False, 'game_mode': None, 'game_type': None, 'cup_type': 'QUICK_MATCH'}, 'players': []}

**Comment by rishubil** [comment]
I'm guessing from the log information that there are two major causes of the problem.

The first reason is that, I don't know if it is a feature of Unity games but, the address of the object we are trying to parse is changed occasionally.
In particular, it is often observed when connecting a gamepad such as an xbox controller.
If the address value of the object actually to be parsed is changed, the observer can not operate normally.
Therefore, to solve this problem, I think there is only way to change the method of guessing and importing the object address, or more accurately understand the memory structure of the game.

The second reason is that, it actually does not matter if we solve the first reason, but when It incorrectly guesses the length of Array and List in `models.py`, the print method will repeat indefinitely.
There are some ways to fix this problem, such as inserting the maximum number of iterations on print method, etc. But, unfortunately, this is a temporary measure of the failure of parsing.

Anyway, this issue is probably not closed and I think it will be here for a long time.
If someone who has experienced game memory parsing can help, there is hope that it can be resolved more quickly.

**Comment by rishubil** [comment]
I'm trying to fix this issue, but not sure it should be fixed.
But, after commit ead6f290f5e4d7fdc6226d613b8cded3c3f6926c, it may be work better.

**Comment by rishubil** [comment]
It may be fixed after v0.1.3 but not tested.
If a similar problem occurs, it will open the issue again.
